src/Docker/Runner/DataLoader/AbstractWorkspaceProvider.php

```diff
@@ -0,0 +1,86 @@
+<?php
+
+namespace Keboola\DockerBundle\Docker\Runner\DataLoader;
+
+use Keboola\DockerBundle\Exception\ApplicationException;
+use Keboola\InputMapping\Staging\ProviderInterface;
+use Keboola\StorageApi\Client;
+use Keboola\StorageApi\Components;
+use Keboola\StorageApi\Workspaces;
+
+abstract class AbstractWorkspaceProvider implements ProviderInterface
+{
+    /** @var Client */
+    protected $client;
+
+    /** @var array */
+    protected $workspace;
+
+    /** @var string */
+    protected $componentId;
+
+    /** @var string */
+    protected $configurationId;
+
+    public function __construct(Client $client, $componentId, $configurationId)
+    {
+        $this->client = $client;
+        $this->componentId = $componentId;
+        $this->configurationId = $configurationId;
+        $this->workspace = null;
+    }
+
+    abstract protected function getType();
+
+    protected function createWorkspace()
+    {
+        if ($this->configurationId) {
+            $components = new Components($this->client);
+            $this->workspace = $components->createConfigurationWorkspace(
+                $this->componentId,
+                $this->configurationId,
+                ['backend' => $this->getType()]
+            );
+        } else {
+            $workspaces = new Workspaces($this->client);
+            $this->workspace = $workspaces->createWorkspace(['backend' => $this->getType()]);
+        }
+    }
+
+    public function getWorkspaceId()
+    {
+        if (!$this->workspace) {
+            $this->createWorkspace();
+        }
+        return $this->workspace['id'];
+    }
+
+    public function cleanup()
+    {
+        if ($this->workspace) {
+            $workspaces = new Workspaces($this->client);
+            $workspaces->deleteWorkspace($this->workspace['id'], ['async' => true]);
+            $this->workspace = null;
+        }
+    }
+
+    public function getCredentials()
+    {
+        if (!$this->workspace) {
+            $this->createWorkspace();
+        }
```

Comment:
I'm not sure if it may be better to throw an error in this case.  imo calling getCredentials on a non-existant workspace should be an error.

Reply:
It's not calling getCredentials on a non-existent workspace, it's calling WorkspaceProvider for the first time and lazy initialization of the workspace. I.e. we don't know which of getWorkspaceId or getCredentials will be called first.